pyro/infer/svgd.py

```diff
@@ -0,0 +1,220 @@
+from pyro import poutine
+from pyro.distributions import Delta
+from pyro.infer.trace_elbo import Trace_ELBO
+from pyro.infer.autoguide.guides import AutoContinuous
+from pyro.infer.autoguide.initialization import init_to_sample
+from pyro.distributions.util import sum_rightmost, copy_docs_from
+
+
+def vectorize(fn, num_particles, max_plate_nesting):
+    def _fn(*args, **kwargs):
+        with pyro.plate("num_particles_vectorized", num_particles, dim=-max_plate_nesting - 1):
+            return fn(*args, **kwargs)
+    return _fn
+
+
+class _SVGDGuide(AutoContinuous):
+    """
+    This modification of :class:`AutoContinuous` is used internally in the
+    :class:`SVGD` inference algorithm.
+    """
+    def __init__(self, model):
+        super(_SVGDGuide, self).__init__(model, prefix="svgd", init_loc_fn=init_to_sample)
+
+    def get_posterior(self, *args, **kwargs):
+        svgd_particles = pyro.param("svgd_particles", self._init_loc)
+        return Delta(svgd_particles, event_dim=1)
+
+
+class SteinKernel(object, metaclass=ABCMeta):
+    """
+    Abstract class for kernels used in the :class:`SVGD` inference algorithm.
+    """
+
+    @abstractmethod
+    def log_kernel_and_grad(self, particles):
+        """
+        Compute the component kernels and their gradients.
+
+        :param particles: a tensor with shape (N, D)
+        :returns: A pair (`log_kernel`, `kernel_grad`) where `log_kernel` is a (N, N, D)-shaped
+            tensor equal to the logarithm of the kernel and `kernel_grad` is a (N, N, D)-shaped
+            tensor where the entry (n, m, d) represents the derivative of `log_kernel` w.r.t.
+            x_{m,d}, where x_{m,d} is the d^th dimension of particle m.
+        """
+        raise NotImplementedError
+
+
+class RBFSteinKernel(SteinKernel):
+    """
+    A RBF kernel for use in the SVGD inference algorithm. The bandwidth of the kernel is chosen from the
+    particles using a simple heuristic as in reference [1].
+
+    References
+
+    [1] "Stein Variational Gradient Descent: A General Purpose Bayesian Inference Algorithm,"
+        Qiang Liu, Dilin Wang
+    """
+    def __init__(self, bandwidth_factor=None):
+        """
+        :param float bandwidth_factor: Optional factor by which to scale the bandwidth
+        """
+        self._bandwidth_factor = bandwidth_factor
+
+    def _bandwidth(self, norm_sq):
+        """
+        Compute the bandwidth along each dimension using the median pairwise squared distance between particles.
+        """
+        num_particles = norm_sq.size(0)
+        index = torch.arange(num_particles)
+        norm_sq = norm_sq[index > index.unsqueeze(-1), ...]
+        median = norm_sq.median(dim=0)[0]
+        if self.bandwidth_factor is not None:
+            median = self.bandwidth_factor * median
+        assert median.shape == norm_sq.shape[-1:]
+        return median / math.log(num_particles + 1)
+
+    @torch.no_grad()
+    @copy_docs_from(SteinKernel.log_kernel_and_grad)
```

Comment:
@neerajprad is this invocation ok?

Reply:
Hmm..I don't think that works at the level of class methods. Why not just use this to annotate the class itself? It will only override for methods that don't have a docstring of their own.

Reply:
thanks, done (i think)